App Config README (#3327)
Azure/azure-sdk-for-java: applicationconfig/client/README.md

# Azure App Configuration client library for Java
Azure App Configuration is a managed service that helps developers centralize their application configurations simply and securely.

Modern programs, especially programs running in a cloud, generally have many components that are distributed in nature. Spreading configuration settings across these components can lead to hard-to-troubleshoot errors during an application deployment. Use App Configuration to store all the settings for your application and secure their accesses in one place.

Use the client library for App Configuration to create and manage application configuration settings.

> [Source code][source_code] | [Package (Maven)][package] | [API reference documentation][api_documentation] | [Product documentation][azconfig_docs]

Reviewer: File an issue here to fix these links when they are known.

Author: https://github.com/Azure/azure-sdk-for-java/issues/3437

## Getting started

### Prerequisites

- [Java Development Kit (JDK)][jdk] with version 8 or above

> - [Apache Maven][maven] 3.0 or above

Reviewer: I wouldn't list Maven as a dependency to using this library - it can be used with other tools or downloaded directly.

You need to install the Azure App Configuration CLI extension first by executing the following command:
```Powershell
az extension add -n appconfig
```

After that, create the Configuration Store:
```Powershell
az appconfig create --name <config-store-name> --resource-group <resource-group-name> --location eastus
```

### Authenticate the client

In order to interact with the App Configuration service you'll need to create an instance of the Configuration Client class. To make this possible you'll need the connection string of the Configuration Store.

#### Get Credentials

Use the [Azure CLI][azure_cli] snippet below to get the connection string from the Configuration Store.

```Powershell
az appconfig credential list --name <config-store-name>
```

Alternatively, get the connection string from the Azure Portal.

#### Create Client

Once you have the value of the connection string you can create the ConfigurationClient:

```java
ConfigurationClient client = ConfigurationClient.builder()
        .credentials(new ConfigurationClientCredentials(connectionString))
        .build();
```

## Key concepts

### Configuration Setting

> Is the fundamental resource within a Configuration Store. In its simplest form it is a key and a value. However, there are additional properties such as the modifiable content type and tags fields that allow the value to be interpreted or associated in different ways.

Reviewer: ```suggestion
A configuration setting is the fundamental resource within a Configuration Store. In its simplest form it is a key and a value. However, there are additional properties such as the modifiable content type and tags fields that allow the value to be interpreted or associated in different ways.
```

The Label property of a Configuration Setting provides a way to separate Configuration Settings into different dimensions. These dimensions are user defined and can take any form. Some common examples of dimensions to use for a label include regions, semantic versions, or environments. Many applications have a required set of configuration keys that have varying values as the application exists across different dimensions. For example, MaxRequests may be 100 in "NorthAmerica", and 200 in "WestEurope". By creating a Configuration Setting named MaxRequests with a label of "NorthAmerica" and another, only with a different value, in the "WestEurope" label, a solution can be achieved that allows the application to seamlessly retrieve Configuration Settings as it runs in these two dimensions.

Properties of a Configuration Setting:

```Java
String key;

String label;

String value;

String contentType;

String etag;

OffsetDateTime lastModified;

boolean locked;

Map<String, String> tags;

> ```

Reviewer: I don't see much value in this code listing

- setSetting creates a setting if it doesn't exist or overrides an existing setting.
```Java
String connectionString = <connection_string>;
ConfigurationClient client = ConfigurationClient.builder()
        .credentials(new ConfigurationClientCredentials(connectionString))
        .build();
ConfigurationSetting setting = client.setSetting("some_key", "some_value");
```

### Retrieve a Configuration Setting

Retrieve a previously stored Configuration Setting by calling getSetting.
```Java
String connectionString = <connection_string>;
ConfigurationClient client = ConfigurationClient.builder()
        .credentials(new ConfigurationClientCredentials(connectionString))
        .build();
client.setSetting("some_key", "some_value");
ConfigurationSetting setting = client.getSetting("some_key");
```

### Update an existing Configuration Setting

Update an existing Configuration Setting by calling updateSetting.
```Java
String connectionString = <connection_string>;
ConfigurationClient client = ConfigurationClient.builder()
        .credentials(new ConfigurationClientCredentials(connectionString))
        .build();
client.setSetting("some_key", "some_value");

> ConfigurationSetting setting = client.updateSetting("some_key", "new)value");

Reviewer 2: NIT: maybe changing `new)value` to `new_value`?

client.setSetting("some_key", "some_value");
ConfigurationSetting setting = client.updateSetting("some_key", "new_value");
```

### Delete a Configuration Setting

Delete an existing Configuration Setting by calling deleteSetting.
```Java
ConfigurationClient client = ConfigurationClient.builder()
        .credentials(new ConfigurationClientCredentials(connectionString))
        .build();
client.setSetting("some_key", "some_value");
ConfigurationSetting setting = client.deleteSetting("some_key");
```

## Troubleshooting

### General

> When you interact with App Configuration using the .NET SDK, errors returned by the service correspond to the same HTTP status codes returned for [REST API][azconfig_rest] requests.

Reviewer: ```suggestion
When you interact with App Configuration using this Java client library, errors returned by the service correspond to the same HTTP status codes returned for [REST API][azconfig_rest] requests.
```

For example, if you try to retrieve a Configuration Setting that doesn't exist in your Configuration Store, a `404` error is returned, indicating `Not Found`.

```Java
ConfigurationClient client = ConfigurationClient.builder()
        .credentials(new ConfigurationClientCredentials(connectionString))
        .build();
ConfigurationSetting setting = client.getSetting("some_key");

> ````

Reviewer: This code doesn't show how to handle exceptions.

> # App Configuration client library for Java

Reviewer 2: According to their docs, the name should be `Azure App Configuration`

Reviewer 2: https://docs.microsoft.com/en-us/azure/azure-app-configuration/

> Azure App Configuration allows developers to store, retrieve and manage access to application settings all in one place. It is easy to set up and simple to use from any application. It gives developers the ability to modify an application's behavior on demand without having to redeploy the application.

Reviewer 2: This doesn't follow the [Readme template](https://github.com/Azure/azure-sdk/blob/master/docs/README-TEMPLATE.md) . The service description should be taken from: https://github.com/Azure/azure-sdk/blob/master/docs/README-TEMPLATE.md

> Azure App Configuration allows developers to store, retrieve and manage access to application settings all in one place. It is easy to set up and simple to use from any application. It gives developers the ability to modify an application's behavior on demand without having to redeploy the application.

Reviewer 2: I am curious to know where this description came from... I will stay away from saying things like `easy` 
=> BTW in .NET we said `...that allows to easily store and manage ...` I got feedback about that word too, so I need to change it.
I was thinking on using the description the service uses... like 

*Azure App Configuration provides a service to centrally manage application settings. Modern programs, especially programs running in a cloud, generally have many components that are distributed in nature. Spreading configuration settings across these components can lead to hard-to-troubleshoot errors during an application deployment. Use App Configuration to store all the settings for your application and secure their accesses in one place.*

What do you think?

Author: It came from the [Overview](https://docs.microsoft.com/en-us/azure/azure-app-configuration/overview) section in the App Configuration docs.

I don't mind the wording change especially if it keeps us consistent across languages.

Reviewer 2: hmm I don't see it there, but either way, I think you should include more info about the service like with what I suggested. 
As it currently stands it is a little too generic? or simple?

> Use the client library for Application Configuration to:

Reviewer 2: There should be consistency when using the name of the service. In some places you use `App` and in others `Application`.

> - Adding a new configuration

Reviewer 2: Consider changing to `Add`, `Update`, `Retrieve`, `Delete`

> - Adding a new configuration

Reviewer 2: ... a new Setting Configuration?

- Updating an existing configuration
- Retrieving a configuration
- Deleting a configuration

[Source code]() | [Package (Maven)]() | [API reference documentation]() | [Product documentation](https://docs.microsoft.com/en-us/azure/azure-app-configuration/)

## Getting started

### Prerequisites

- JDK 8 or later

> - Maven

Reviewer 2: link? is there a version that is required? or a minimum?

> - Azure Service Principal

Reviewer 2: Not sure what you mean with this...
But take a look at: https://docs.microsoft.com/en-us/azure/azure-app-configuration/quickstart-aspnet-core-app#prerequisites

Something like that, plus whatever is needed for Java.

Reviewer 2: still missing that an Azure subscription is needed. See the link I am providing where they list the prerequisites

> - App Configuration resource

Reviewer 2: you mean an App Configuration Store?

### Adding the package to your product

```xml
<dependency>
    <groupId>com.azure</groupId>
    <artifactId>azure-app-configuration</artifactId>
    <version>1.0.0-SNAPSHOT</version>
</dependency>
```

> 

Reviewer 2: This document is missing https://github.com/Azure/azure-sdk/blob/master/docs/README-TEMPLATE.md#authenticate-the-client

When a configuration is modified a point-in-time snapshot is recorded, snapshots are retained for seven days.

### Responses

```Java
public class ConfigurationSetting {
    private String key;
    private String label;
    private String value;
    private String contentType;
    private String etag;
    private OffsetDateTime lastModified;
    private boolean locked;
    private Map<String, String> tags;
}
```

Successful requests to the service will have a response containing a ConfigurationSetting as the body. ConfigurationSetting contains the key-value and metadata about the configuration.

An unsuccessful request will throw a ServiceRequestException, in the exception response there will be a HTTP status code that indicates why the request failed.

For further information about responses view the [Rest API reference](https://github.com/Azure/AppConfiguration#rest-api-reference).

## Examples

### Creating a new configuration

### Updating an existing configuration

### Retrieving a configuration

### Deleting a configuration

## Troubleshooting

### Unauthorized Response (401)

> Making requests with an unauthenicate client results in the service returning an unauthorized response (401), verify that the credentials used in the construction of the client are correct.

Reviewer: ```suggestion
Making requests with an unauthenticated client results in the service returning an unauthorized response (401), verify that the credentials used in the construction of the client are correct.
```

# Azure App Configuration client library for Java
Azure App Configuration provides a service to centrally manage application settings. Modern programs, especially programs running in a cloud, generally have many components that are distributed in nature. Spreading configuration settings across these components can lead to hard-to-troubleshoot errors during an application deployment. Use App Configuration to store all the settings for your application and secure their accesses in one place.

Use the client library for App Configuration to:
- Add a new Configuration Setting
- Update an existing Configuration Setting
- Retrieve a Configuration Setting
- Delete a Configuration Setting

[Source code]() | [Package (Maven)]() | [API reference documentation]() | [Product documentation](https://docs.microsoft.com/en-us/azure/azure-app-configuration/)

## Getting started

### Prerequisites

- [Java Development Kit (JDK)](https://docs.microsoft.com/en-us/java/azure/java-supported-jdk-runtime?view=azure-java-stable) with version 8 or above
- [Apache Maven](https://maven.apache.org/) 3.0 or above

> - Azure Service Principal

Reviewer 3: Do we need a service principal to use this SDK? I thought we just used a connection string for the service.

Author: Removed

# Azure App Configuration client library for Java
Azure App Configuration is a managed service that helps developers centralize their application configurations simply and securely.

Use the client library for App Configuration to:
- Add a new setting configuration
- Update an existing setting configuration
- Retrieve a setting configuration
- Delete a setting configuration

[Source code]() | [Package (Maven)]() | [API reference documentation]() | [Product documentation](https://docs.microsoft.com/en-us/azure/azure-app-configuration/)

## Getting started

> ### Prerequisites

Reviewer 2: having links for each of this would be useful...

- JDK 8 or later
- Apache Maven 3 or above
- Azure Service Principal
- App Configuration Store

### Adding the package to your product

```xml
<dependency>
    <groupId>com.azure</groupId>
    <artifactId>azure-app-configuration</artifactId>
    <version>1.0.0-SNAPSHOT</version>
</dependency>
```

### Authenticate the client

```java
String connectionString = "endpoint={endpoint_value};id={id_value};secret={secret_value}";
ConfigurationClient client = ConfigurationClient.builder()
        .credentials(new ConfigurationClientCredentials(connectionString)
        .build();
```

> Using the Azure web portal to open your App Configuration Store's Access Key section, choose one of the two connection strings and use it to replace the string value above.

Reviewer 2: Should we add an option for Azure CLI?

> ## Key concepts

Reviewer 2: Although these are the key concepts for the service team, we should focus on the key concepts for people understanding the service and our SDK. Point in time doesn't seem that relevant for example

### [Key-value store](https://docs.microsoft.com/en-us/azure/azure-app-configuration/concept-key-value)

> Configurations are stored using key-value pairs. Keys must be unique for each configuration so it's common practice to organize keys into hierarchical namespaces using delimiting characters, for example AppName:Service1:DB:Endpoint. Optionally a label attribute may be given to a key to differentiate two configurations, for example having different versions of the configuration. Values are stored as unicode strings.

Reviewer 2: I know you are copying this from the Service team, but in the SDK we are referring to the objects as `Configuration Setting`. Might as well address this as "Configuration Settings are stored..."

> Configurations are stored using key-value pairs. Keys must be unique for each configuration so it's common practice to organize keys into hierarchical namespaces using delimiting characters, for example AppName:Service1:DB:Endpoint. Optionally a label attribute may be given to a key to differentiate two configurations, for example having different versions of the configuration. Values are stored as unicode strings.

Reviewer 2: I'm not a fan of this description. It misleads on how to set keys. For example, you are [omitting ](App Configuration treats keys as a whole. It doesn't parse keys to figure out how their names are structured or enforce any rule on them.) that the service doesn't do any parsing and therefore, the customer will have to do it.

Also, you are referring to what a key is, but the title talks about a key-value store, which in our SDK language means a Configuration Setting, which includes way more information than a key... 
Something like [this](https://github.com/Azure/azure-sdk-for-net/tree/master/src/SDKs/Azure.ApplicationModel.Configuration/data-plane#configuration-setting) might be better

### [Point-in-time snapshot](https://docs.microsoft.com/en-us/azure/azure-app-configuration/concept-point-time-snapshot)

When a configuration is modified a point-in-time snapshot is recorded, snapshots are retained for seven days.

### Responses

```Java
public class ConfigurationSetting {
    private String key;
    private String label;
    private String value;
    private String contentType;
    private String etag;
    private OffsetDateTime lastModified;
    private boolean locked;
    private Map<String, String> tags;
}
```

Successful requests to the service will have a response containing a ConfigurationSetting as the body. ConfigurationSetting contains the key-value and metadata about the configuration.

An unsuccessful request will throw an error that indicates why the request failed. (todo - update this with further error type explanation when implemented)

For further information about responses view the [Rest API reference](https://github.com/Azure/AppConfiguration#rest-api-reference).

## Examples

> ### Create a new setting configuration

Reviewer 2: Is the idea to add these examples in future PR?

Author: Yup, @conniey is creating samples in PR #3288, after that is merged in we can take some of the code snippets from it and put them in the README

Reviewer 2: Ohh right!

### Update an existing setting configuration

### Retrieve a setting configuration

### Delete a setting configuration

## Troubleshooting

### *TODO - Add Exception Response Type*

Making requests with an unauthenticated client results in the service returning an unauthorized response (401), verify that the credentials used in the construction of the client are correct.

### *TODO - Add Exception Response Type*

> Attempting to update or delete a locked configuration results in a conflict response (409), unlock the configuration to resolve this issue.

Reviewer 2: Lock and unlock functionality has been deleted for the API. See PR https://github.com/Azure/azure-sdk-for-java/pull/3319

Author: I feel indifferent about removing this, while the functionality was removed from the API users could still run into locked setting configurations which would result in the error response.

If we believe that it is uncommonly for setting configurations to be locked I will go ahead and remove mention of them from this section.

Reviewer 2: how will they lock a setting?

Reviewer 2: Removing this functionality from the API forces customers to lock/unlock only by REST calls. Notice that the Azure Portal doesn't support lock/unlock either.

It is an uncommon scenario

Author: The portal does allow for locking/unlocking but I agree it should be relatively uncommon, will remove the section.

![image](https://user-images.githubusercontent.com/48699787/55837405-26162f00-5ad6-11e9-89a6-cd0a608a493c.png)

Reviewer 2: Ohh they added it. Cool :)
